File: WEEK_4_C/Q3.c
#include <stdio.h>
#include <stdlib.h>

struct Node{ int data; struct Node *prev,*next; };

int main(){
    struct Node *head=NULL,*last=NULL,*t;
    for(int i=1;i<=3;i++){
        t=(struct Node*)malloc(sizeof(struct Node));
        t->data=i; t->next=NULL; t->prev=last;
        if(!head) head=t; else last->next=t;
        last=t;
    }
    for(t=head;t;t=t->next) printf("%d ",t->data);
    return 0;
}

File: WEEK_4_Python/Q3.py
class DNode:
    def __init__(self, data):
        self.data, self.prev, self.next = data, None, None

def print_dll(head):
    cur = head
    while cur:
        print(cur.data, end=" ")
        cur = cur.next
    print()

def reverse_dll(head):
    cur = head
    prev = None
    while cur:
        cur.prev, cur.next = cur.next, cur.prev
        prev = cur
        cur = cur.prev
    return prev

# Example
head = DNode(1); head.next = DNode(2); head.next.prev = head; head.next.next = DNode(3); head.next.next.prev = head.next
head = reverse_dll(head)
print_dll(head)   
